### PR TITLE
feat(APIG-CPC 48): Get billing info by ID 

### PR DESCRIPTION
**Pull request by Ludo67**

### CODE REVIEW

**Review by VladPopa616 — CHANGES_REQUESTED** [review]
The reason your build fails is that you have a @NoArgConstructor annotation. Remove that and try again.

**Comment by Ludo67** [comment]
Will do asap

**Comment by Ludo67** [comment]
Screwed up the merge conflict sorry. Will do it as soon as I get home

File: api-gateway/src/main/java/com/petclinic/bffapigateway/domainclientlayer/BillServiceClient.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    public Mono<BillDetails> getBilling(final int billId) {
+        return webClientBuilder.build().get()
+                .uri(billServiceUrl + "/{billId}", billId)
+                .retrieve()
+                .bodyToMono(BillDetails.class);
+    }
+}
+
+
```
Comment (VladPopa616) [inline api-gateway/src/main/java/com/petclinic/bffapigateway/domainclientlayer/BillServiceClient.java:38]: This seems fine.

File: api-gateway/src/main/java/com/petclinic/bffapigateway/dtos/BillDetails.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+@Data
+public class BillDetails {
+    private int billId;
+    private Date date;
+    private int customerId;
+    private String visitType;
+    private double amount;
+
+}
```
Comment (VladPopa616) [inline api-gateway/src/main/java/com/petclinic/bffapigateway/dtos/BillDetails.java:21]: DTO class, nothing to say here.

File: api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java [hunk 3]
```diff
@@ -50,8 +59,13 @@
 
     private final VetsServiceClient vetsServiceClient;
 
+
     private final AuthServiceClient authServiceClient;
 
+    private final BillServiceClient billServiceClient;
+
+
+
```
Comment (VladPopa616) [inline api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java:68]: Watch the spacing

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 4]
```diff
@@ -206,38 +243,74 @@ void deleteUser() {
...
+                .expectBody()
+                .jsonPath("$.billId").isEqualTo(1)
+                .jsonPath("$.customerId").isEqualTo(entity.getCustomerId())
+                .jsonPath("$.visitType").isEqualTo(entity.getVisitType())
+                .jsonPath("$.amount").isEqualTo(entity.getAmount());
+
+
+
+
+        assertEquals(entity.getBillId(), 1);
```
Comment (LawrenceLiang02) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:282]: This is questionable, you are just getting what you have set and comparing it but ok.

File: api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java [hunk 5]
```diff
@@ -60,6 +74,14 @@
...
                 );
     }
+//check this
+    @GetMapping(value = "bills/{billId}")
+    public Mono<BillDetails> getBillingInfo(final @PathVariable int billId)
+    {
+        return billServiceClient.getBilling(billId);
+    }
+
+
```
Comment (VladPopa616) [inline api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java:84]: GetMapping seems good here

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 6]
```diff
@@ -206,38 +243,74 @@ void deleteUser() {
...
+
+
+    }
+
+
+    @Test
+    void getBillingByRequestMissingPath(){
```
Comment (LawrenceLiang02) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:289]: Do you think the tests should start with "should" as explained in class?

File: api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java [hunk 7]
```diff
@@ -121,10 +144,10 @@
...
 
 
-          return customersServiceClient.updateOwner(od,ownerId)
-                    .flatMap(owner ->
-                            visitsServiceClient.getVisitsForPets(owner.getPetIds())
-                                    .map(addVisitsToOwner(owner)));
+        return customersServiceClient.updateOwner(od,ownerId)
+                .flatMap(owner ->
+                        visitsServiceClient.getVisitsForPets(owner.getPetIds())
+                                .map(addVisitsToOwner(owner)));
```
Comment (VladPopa616) [inline api-gateway/src/main/java/com/petclinic/bffapigateway/presentationlayer/BFFApiGatewayController.java:150]: So what exactly was the change here? I'm assuming indents?

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 8]
```diff
@@ -117,7 +148,9 @@ void getUserDetails() {
 
         assertEquals(user.getId(), 1);
     }
-  
+
+
+
```
Comment (VladPopa616) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:153]: Big empty space.

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 9]
```diff
@@ -206,38 +243,74 @@ void deleteUser() {
...
-//        assertEquals(od.getTelephone(),"5553451125");
-//
-//
-//    }
+
+
+    //private static final int BILL_ID = 1;
+
+    @Test
+    public void getBillById(){
```
Comment (VladPopa616) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:251]: Remember Jason's naming convention for tests it should start with "should" or "test" or something like that. Applies to all tests.

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 10]
```diff
@@ -206,38 +243,74 @@ void deleteUser() {
...
+                .jsonPath("$.amount").isEqualTo(entity.getAmount());
+
+
+
+
+        assertEquals(entity.getBillId(), 1);
+
+
+    }
+
```
Comment (VladPopa616) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:286]: Overall good test.

File: api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java [hunk 11]
```diff
@@ -206,38 +243,74 @@ void deleteUser() {
...
+                .accept(APPLICATION_JSON)
+                .exchange()
+                .expectStatus().isNotFound()
+                .expectHeader().contentType(APPLICATION_JSON)
+                .expectBody()
+                .jsonPath("$.path").isEqualTo("/bills/100")
+                .jsonPath("$.message").isEqualTo(null);
+
+    }
+
```
Comment (VladPopa616) [inline api-gateway/src/test/java/com/petclinic/bffapigateway/presentationlayer/ApiGatewayControllerTest.java:313]: Good error case tests.

**Review by VladPopa616 — APPROVED** [review]
Some formatting issues, but I will approve it.

**Pull request merged**